plugins/API/API.php

```diff
@@ -520,12 +532,16 @@ public function getSuggestedValuesForSegment($segmentName, $idSite)
             throw new \Exception("There was no data to suggest for $segmentName");
         }
 
-        // Cleanup data to return the top suggested (non empty) labels for this segment
-        $values = $table->getColumn($segmentName);
+        if ($suggestedValuesCallbackRequiresTable) {
+            $values = call_user_func($segmentFound['suggestedValuesCallback'], $idSite, $maxSuggestionsToReturn, $table);
+        } else {
+            // Cleanup data to return the top suggested (non empty) labels for this segment
+            $values = $table->getColumn($segmentName);
 
-        // Select also flattened keys (custom variables "page" scope, page URLs for one visit, page titles for one visit)
-        $valuesBis = $table->getColumnsStartingWith($segmentName . ColumnDelete::APPEND_TO_COLUMN_NAME_TO_KEEP);
-        $values = array_merge($values, $valuesBis);
+            // Select also flattened keys (custom variables "page" scope, page URLs for one visit, page titles for one visit)
+            $valuesBis = $table->getColumnsStartingWith($segmentName . ColumnDelete::APPEND_TO_COLUMN_NAME_TO_KEEP);
+            $values = array_merge($values, $valuesBis);
+        }
```

Comment:
I noticed many methods needing refactoring :) I even added a TODO for the first one. Then I saw the rest and lost hope :)